src/utl/src/timer.cpp

```diff
@@ -74,4 +76,23 @@ DebugScopedTimer::~DebugScopedTimer()
   debugPrint(logger_, tool_, group_.c_str(), level_, msg_, *this);
 }
 
+ScopedStatistics::ScopedStatistics(utl::Logger* logger)
+    : Timer(), logger_(logger)
+{
+}
+
+size_t ScopedStatistics::getPeakMemoryUsage()
+{
+  struct rusage rusage;
+  getrusage(RUSAGE_SELF, &rusage);
```

Comment:
error handling for getrusage missing. There may not be access rights to get this information or otherwise it is not available?

Reply:
Actually, what I want to know is maximum virtual memory, not maximum resident memory set. The only reason why ORFS is settling for maximum resident memory set is because `ps` doesn't measure maximum virtual memory.

A discussion on cores and max virtual memory  vs. maximum resident memory set.

Reply:
If you can't measure it you can't report it so I'm not sure what is requested here.

I would argue that max virtual memory is not very interesting in 64bit computing.  You'll never run out of virtual address space and thrashing is a resident set size issue.

Reply:
max virtual address space can be measured in C++ easily enough, it just the `time` command that doesn't support it, which makes it awkward for ORFS to report it.

I disagree that virtual memory size isn't interesting. resident set size will always be smaller than physical memory and and doesn't give any indication that you have too large a working set and that the system could be thrashing. That said, even if max virtual memory is larger than physical memory, doesn't mean that there will be thrashing.

Reply:
How is it easily measured?

rss ~= physical memory is a warning sign.  vsz > physical memory doesn't say much.

Reply:
I don't have a mac, but when I compile and run this on Ubuntu, I get the output below:

```C++
#include <iostream>
#include <unistd.h>

#ifdef __linux__
#include <fstream>
#elif __APPLE__
#include <mach/mach.h>
#endif

void getMemoryUsage(size_t &rss, size_t &vsz) {
#ifdef __linux__
    std::ifstream statm("/proc/self/statm");
    if (!statm.is_open()) {
        std::cerr << "Error opening /proc/self/statm" << std::endl;
        return;
    }

    size_t pageSize = sysconf(_SC_PAGESIZE);
    size_t size, resident, share, text, lib, data, dt;

    statm >> size >> resident >> share >> text >> lib >> data >> dt;

    rss = resident * pageSize; // Convert from pages to bytes
    vsz = size * pageSize;     // Convert from pages to bytes

    statm.close();
#elif __APPLE__
    mach_task_basic_info info;
    mach_msg_type_number_t infoCount = MACH_TASK_BASIC_INFO_COUNT;

    if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO, (task_info_t)&info, &infoCount) != KERN_SUCCESS) {
        std::cerr << "Error getting task info" << std::endl;
        return;
    }

    rss = info.resident_size;
    vsz = info.virtual_size;
#endif
}

int main() {
    size_t rss = 0, vsz = 0;
    getMemoryUsage(rss, vsz);
    std::cout << "RSS: " << rss << " bytes" << std::endl;
    std::cout << "VSZ: " << vsz << " bytes" << std::endl;
    return 0;
}
```

Output:

```
$ ./a.out
RSS: 3440640 bytes
VSZ: 6377472 bytes
```

Reply:
Would it be more convenient to have both RSS and VSZ information?

Reply:
For my part, I think we need both. Over time, we may develop better understanding of how they interact with OpenROAD and decide we only need rsz or rss and why only one is sufficient, but for now I'm definitely missing rsz when analyzing how much more resources I need to give a build server...